src/Forms/SearchableDropdownTrait.php

```diff
@@ -0,0 +1,594 @@
+     *
+     * Calling this will also call setUseSearchContext(true)
+     */
+    public function setSearchContext(?SearchContext $searchContext): static
+    {
+        $this->searchContext = $searchContext;
+        $this->setUseSearchContext(true);
+        return $this;
+    }
+
+    /**
+     * Get whether to use a dynamic placeholder if a normal placeholder is not set
+     */
+    public function getUseDynamicPlaceholder(): bool
+    {
+        return $this->useDynamicPlaceholder;
+    }
+
+    /**
+     * Set whether to use a dynamic placeholder if a normal placeholder is not set
+     */
+    public function setUseDynamicPlaceholder(bool $useDynamicPlaceholder): static
+    {
+        $this->useDynamicPlaceholder = $useDynamicPlaceholder;
+        return $this;
+    }
+
+    /**
+     * Get whether to use a search context instead searching on labelField
+     */
+    public function getUseSearchContext(): bool
+    {
+        return $this->useSearchContext;
+    }
+
+    /**
+     * Set whether to use a search context instead searching on labelField
+     */
+    public function setUseSearchContext(bool $useSearchContext): static
+    {
+        $this->useSearchContext = $useSearchContext;
+        return $this;
+    }
+
+    /**
+     * Get whether the field allows searching by typing characters into field
+     */
+    public function getIsSearchable(): bool
+    {
+        return $this->isSearchable;
+    }
+
+    /**
+     * Set whether the field allows searching by typing characters into field
+     */
+    public function setIsSearchable(bool $isSearchable): static
+    {
+        $this->isSearchable = $isSearchable;
+        return $this;
+    }
+
+    /**
+     * This returns an array rather than a DataList purely to retain compatibility with ancestor getSource()
+     */
+    public function getSource(): array
+    {
+        return $this->getListMap($this->sourceList);
+    }
+
+    /*
+     * @param mixed $source
+     */
+    public function setSource($source): static
+    {
+        if (!is_a($source, DataList::class)) {
+            throw new InvalidArgumentException('Source must be a DataList');
+        }
+        // Setting to $this->sourceList instead of $this->source because SelectField.source
+        // docblock type is array|ArrayAccess i.e. does not allow DataList
+        $this->sourceList = $source;
+        return $this;
+    }
+
+    /**
+     * Get the field to use for the label of the option
+     *
+     * The default value of 'Title' will map to DataObject::getTitle() if a Title DB field does not exist
+     */
+    public function getLabelField(): string
+    {
+        return $this->labelField;
+    }
+
+    /**
+     * Set the field to use for the label of the option
+     */
+    public function setLabelField(string $labelField): static
+    {
+        $this->labelField = $labelField;
+        return $this;
+    }
+
+    public function getAttributes(): array
```

Comment:
```suggestion
    protected function getDefaultAttributes(): array
```
This seems to be by far the more common one to override - `getAttributes()` calls `getDefaultAttributes()` - and then it provides an extension hook. So we should override `getDefaultAttributes()`, and let `getAttributes()` do its thing.

Reply:
It's not more common?. Many FormField subclasses implement getAttributes().  getDefaultAttributes() is defined on FormField but not subclasses.  I will leave this as is.

Reply:
If you do want to keep overriding `getAttributes()` instead of `getDefaultAttributes()`, please re-implement the extension hook that `FormField::getAttributes()` applies.

Reply:
We're calling `parent::getAttributes()` here, which will call `FormField::getAttributes()` and thus its extension

I think the only 'in-between' class that implements `getAttributes()` is `SelectField::getAttributes()` and that also calls `parent::getAttributes()`

Reply:
Ugh, I wish the parent class didn't do that. It means peoples' extensions won't have the full context when they want to edit the attributes. One day I want to look into a better way to declare "have a hook that runs before/after this method".

For now, if this is what we're going with, I guess it has a precedent even if IMO it's a bad one so we can leave as is.